src/value/owned.rs

```diff
@@ -1,186 +1,183 @@
+    fn get(&self, index: usize) -> Option<&Element>;
+    fn len(&self) -> usize;
+    fn is_empty(&self) -> bool {
+        self.len() == 0
     }
+}
+
+/// An in-memory representation of an Ion list
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct List {
+    children: Vec<Element>,
+}
 
-    fn new_f64(float: f64) -> Element {
-        Value::Float(float).into()
+impl List {
+    pub(crate) fn new(children: Vec<Element>) -> Self {
+        Self { children }
     }
 
-    fn new_clob(bytes: &[u8]) -> Element {
-        Value::Clob(bytes.into()).into()
+    pub fn builder() -> ListBuilder {
+        ListBuilder::new()
     }
 
-    fn new_blob(bytes: &[u8]) -> Element {
-        Value::Blob(bytes.into()).into()
+    pub fn clone_builder(&self) -> ListBuilder {
+        ListBuilder::with_initial_elements(&self.children)
     }
+}
 
-    fn new_list<I: IntoIterator<Item = Element>>(seq: I) -> Element {
-        Value::List(seq.into_iter().collect()).into()
+impl IonSequence for List {
+    fn iter(&self) -> ElementsIterator<'_> {
+        ElementsIterator::new(&self.children)
     }
 
-    fn new_sexp<I: IntoIterator<Item = Element>>(seq: I) -> Element {
-        Value::SExpression(seq.into_iter().collect()).into()
+    fn get(&self, index: usize) -> Option<&Element> {
+        self.children.get(index)
     }
 
-    fn new_struct<K: Into<Symbol>, V: Into<Element>, I: IntoIterator<Item = (K, V)>>(
-        structure: I,
-    ) -> Element {
-        Value::Struct(structure.into_iter().collect()).into()
+    fn len(&self) -> usize {
+        self.children.len()
     }
-}
 
-/// An owned implementation of [`IonSequence`]
-#[derive(Debug, Clone)]
-pub struct Sequence {
-    // TODO: Since we've moved the elements Vec to the heap, we could consider replacing it with a
-    //       SmallVec that can store some number of elements (4? 8?) inline. We'd need to benchmark.
-    children: Rc<Vec<Element>>,
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
 }
 
-impl Sequence {
-    pub fn new(children: Vec<Element>) -> Self {
-        Self {
-            children: Rc::new(children),
+impl<S: IonSequence> IonEq for S {
+    fn ion_eq(&self, other: &Self) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+        for (item1, item2) in self.iter().zip(other.iter()) {
+            if !item1.ion_eq(item2) {
+                return false;
+            }
         }
+        true
     }
 }
 
-impl FromIterator<Element> for Sequence {
-    /// Returns an owned sequence from the given iterator of elements.
-    fn from_iter<I: IntoIterator<Item = Element>>(iter: I) -> Self {
-        let mut children: Vec<Element> = Vec::new();
-        for elem in iter {
-            children.push(elem);
-        }
-        Self {
-            children: Rc::new(children),
-        }
+/// An in-memory representation of an Ion s-expression
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SExp {
```

Comment:
I've been pushing for `SExpression` everywhere and this was the PR that made me swing the other way. X_X I started appreciating how short `SExp` is and, given that it's a common abbreviation, began to prefer it.

How would you feel about me renaming the enum variant `SExp` instead of renaming the `struct`?

Reply:
Sure, as long as they're consistent.